Regression in 0.6.6 - panic: runtime error: slice bounds out of range

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behaviour
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->

Deployment as per usual

## Current Behaviour
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->


I'm trying to deploy an image to my local machine but got this error:

```
mini:sample-functions alex$ faas deploy -f dockerhub-stats.yml
Deploying: dockerhub-stats.
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/openfaas/faas-cli/commands.getRegistryAuth(0xc4200ec680, 0xc420013010, 0xf, 0x1, 0x1)
        /go/src/github.com/openfaas/faas-cli/commands/deploy.go:438 +0x1df
github.com/openfaas/faas-cli/commands.runDeployCommand(0xc4200ec640, 0x0, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x162ada0, ...)
        /go/src/github.com/openfaas/faas-cli/commands/deploy.go:168 +0x45c
github.com/openfaas/faas-cli/commands.runDeploy(0x1605a20, 0xc4200ec640, 0x0, 0x2, 0x0, 0x0)
        /go/src/github.com/openfaas/faas-cli/commands/deploy.go:110 +0xcd
github.com/openfaas/faas-cli/vendor/github.com/spf13/cobra.(*Command).execute(0x1605a20, 0xc4200ec600, 0x2, 0x2, 0x1605a20, 0xc4200ec600)
        /go/src/github.com/openfaas/faas-cli/vendor/github.com/spf13/cobra/command.go:698 +0x47a
github.com/openfaas/faas-cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x16053c0, 0xc42007bc70, 0x0, 0x0)
        /go/src/github.com/openfaas/faas-cli/vendor/github.com/spf13/cobra/command.go:783 +0x30e
github.com/openfaas/faas-cli/vendor/github.com/spf13/cobra.(*Command).Execute(0x16053c0, 0x0, 0x0)
        /go/src/github.com/openfaas/faas-cli/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/openfaas/faas-cli/commands.Execute(0xc42000e780, 0x4, 0x4)
        /go/src/github.com/openfaas/faas-cli/commands/faas.go:71 +0xb6
main.main()
        /go/src/github.com/openfaas/faas-cli/app.go:19 +0x7d
```

## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug with details on how you -->
<!--- arrived at your diagnosis, or ideas how to implement the addition or change -->

@BurtonR @nenadilic84 to investigate how this issue could have occurred on Docker for Mac since it is the primary platform for our users.

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. Create function with Go template
2. Build
3. Deploy

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* FaaS-CLI version ( Full output from: `faas-cli version` ):

0.6.6

* Docker version ( Full output from: `docker version` ):

```
Client:
 Version:       18.04.0-ce
 API version:   1.37
 Go version:    go1.9.4
 Git commit:    3d479c0
 Built: Tue Apr 10 18:13:16 2018
 OS/Arch:       darwin/amd64
 Experimental:  true
 Orchestrator:  swarm
```

* Are you using Docker Swarm (FaaS-swarm ) or Kubernetes (FaaS-netes)?

Swarm

* Operating System and version (e.g. Linux, Windows, MacOS):

MacOS Sierra

--- a/README.md
+++ b/README.md
@@ -75,6 +75,7 @@ scoop install faas-cli
 Note: The `scoop` release may not run the latest minor release but is updated regularly.
 
 #### Build from source
+
 > the [contributing guide](CONTRIBUTING.md) has instructions for building from source and for configuring a Golang development environment.
 
 ### Run the CLI
@@ -156,6 +157,17 @@ Specify `lang: Dockerfile` if you want the faas-cli to execute a build or `skip_
 
 Read the blog post/tutorial: [Turn Any CLI into a Function with OpenFaaS](https://blog.alexellis.io/cli-functions-with-openfaas/)
 
+### Private registries
+
+* For Kubernetes
+
+Create a named image pull secret and add the secret name to the `secrets` section of your YAML file or your deployment arguments with `--secret`.
+
+Alternatively you can assign a secret to the node to allow it to pull from your private registry. In this case you do not need to assign the secret to your function.
+
+* For Docker Swarm
+
+For Docker Swarm use the `--send-registry-auth` flag or its shorthand `-a` which will look up your registry credentials in your local credentials store and then transmit them over the wire to the deploy command on the API Gateway. Make sure HTTPS/TLS is enabled before attempting this.
 
 ### Use a YAML stack file
 

--- a/commands/deploy.go
+++ b/commands/deploy.go
@@ -24,12 +24,13 @@ import (
 
 // DeployFlags holds flags that are to be added to commands.
 type DeployFlags struct {
-	envvarOpts  []string
-	replace     bool
-	update      bool
-	constraints []string
-	secrets     []string
-	labelOpts   []string
+	envvarOpts       []string
+	replace          bool
+	update           bool
+	constraints      []string
+	secrets          []string
+	labelOpts        []string
+	sendRegistryAuth bool
 }
 
 var deployFlags DeployFlags
@@ -54,6 +55,8 @@ func init() {
 
 	deployCmd.Flags().StringArrayVar(&deployFlags.constraints, "constraint", []string{}, "Apply a constraint to the function")
 	deployCmd.Flags().StringArrayVar(&deployFlags.secrets, "secret", []string{}, "Give the function access to a secure secret")
+
+	deployCmd.Flags().BoolVarP(&deployFlags.sendRegistryAuth, "send-registry-auth", "a", false, "send registryAuth from Docker credentials manager with the request")
 
 	// Set bash-completion.
 	_ = deployCmd.Flags().SetAnnotation("handler", cobra.BashCompSubdirsInDir, []string{})
@@ -165,7 +168,9 @@ func runDeployCommand(args []string, image string, fprocess string, functionName
 				deployFlags.secrets = mergeSlice(function.Secrets, deployFlags.secrets)
 			}
 
-			function.RegistryAuth = getRegistryAuth(&dockerConfig, function.Image)
+			if deployFlags.sendRegistryAuth {
+				function.RegistryAuth = getRegistryAuth(&dockerConfig, function.Image)
+			}
 
 			fileEnvironment, err := readFiles(function.EnvironmentFile)
 			if err != nil {
@@ -212,8 +217,11 @@ Error: %s`, fprocessErr.Error())
 			return fmt.Errorf("To deploy a function give --yaml/-f or a --image flag")
 		}
 
-		gateway = getGatewayURL(gateway, defaultGateway, gateway, os.Getenv(openFaaSURLEnvironment))
-		registryAuth := getRegistryAuth(&dockerConfig, image)
+		var registryAuth string
+		if deployFlags.sendRegistryAuth {
+			gateway = getGatewayURL(gateway, defaultGateway, gateway, os.Getenv(openFaaSURLEnvironment))
+			registryAuth = getRegistryAuth(&dockerConfig, image)
+		}
 
 		if err := deployImage(image, fprocess, functionName, registryAuth, deployFlags); err != nil {
 			return err
@@ -231,18 +239,25 @@ func deployImage(
 	registryAuth string,
 	deployFlags DeployFlags,
 ) error {
+
 	envvars, err := parseMap(deployFlags.envvarOpts, "env")
+
 	if err != nil {
 		return fmt.Errorf("error parsing envvars: %v", err)
 	}
 
 	labelMap, labelErr := parseMap(deployFlags.labelOpts, "label")
+
 	if labelErr != nil {
 		return fmt.Errorf("error parsing labels: %v", labelErr)
 	}
 
 	functionResourceRequest1 := proxy.FunctionResourceRequest{}
-	proxy.DeployFunction(fprocess, gateway, functionName, registryAuth, image, language, deployFlags.replace, envvars, network, deployFlags.constraints, deployFlags.update, deployFlags.secrets, labelMap, functionResourceRequest1)
+	proxy.DeployFunction(fprocess, gateway, functionName,
+		registryAuth, image, language,
+		deployFlags.replace, envvars, network,
+		deployFlags.constraints, deployFlags.update, deployFlags.secrets,
+		labelMap, functionResourceRequest1)
 
 	return nil
 }

